### PR TITLE
Deprecate GC_TRASH_IGNORE, add tests to confirm Trash works

### PR DESCRIPTION
**Pull request by dlmarion**
This commit:
  1. Deprecated GC_TRASH_IGNORE
  2. Modifies MiniAccumuloCluster so that the user can provide Hadoop properties for the Hadoop configuration that is used by the MiniDFS cluster.
  3. Adds ITs that confirm:
    a. Deleted files do not get put into the Trash in the default configuration (Accumulo enabled / Hadoop disabled) 
because the default value for fs.trash.interval is zero, which means disabled.
    b. Deleted files do get put into the Trash when the default Accumulo configuration is used (GC_TRASH_IGNORE = false) and fs.trash.interval is set to a nonzero value. 
    c. Deleted files do not get put into the Trash when GC_TRASH_IGNORE is set to true and fs.trash.interval is set to a nonzero value. 
    d. Deleted files do not get put into the Trash when Accumulo and Hadoop are enabled, but a custom Trash policy is used that filters file names.

### CODE REVIEW

**Comment by dlmarion** [comment]
Kicked off full IT build

**Comment by dlmarion** [comment]
Full IT build successful

File: test/src/main/java/org/apache/accumulo/test/functional/GarbageCollectorTrashEnabledCustomPolicyIT.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+import org.junit.jupiter.api.Test;
+
+public class GarbageCollectorTrashEnabledCustomPolicyIT extends GarbageCollectorTrashBase {
+
+  public static class NoFlushFilesInTrashPolicy extends TrashPolicyDefault {
+
+    @Override
+    public boolean moveToTrash(Path path) throws IOException {
+      // Don't put flush files in the Trash
+      if (!path.getName().startsWith("F")) {
```
Comment (keith-turner) [inline test/src/main/java/org/apache/accumulo/test/functional/GarbageCollectorTrashEnabledCustomPolicyIT.java:51]: The test could be restructured a bit so that there are F and C or A files to GC.  Then we could got more positive confirmation in the test that this class is being called (we could have zero files in trash for other reasons than this class being called).  Thinking of something like the following.

 1. ingest data
 2. flush table creating a F file
 3. compact table creating an A file and marking the F file for GC
 4. compact table again creating another A file and marking prev A file for GC
 5. wait for the F and A files marked for GC to be gone
 6. assert trash has one file, should be the A file

File: test/src/main/java/org/apache/accumulo/test/functional/GarbageCollectorTrashEnabledIT.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  public void testTrashHadoopEnabledAccumuloEnabled() throws Exception {
+    String table = this.getUniqueNames(1)[0];
+    final FileSystem fs = super.getCluster().getFileSystem();
+    super.makeTrashDir(fs);
+    ArrayList<StoredTabletFile> files = null;
+    TableId tid = null;
+    try (AccumuloClient c = Accumulo.newClient().from(getClientProperties()).build()) {
+      files = super.loadData(super.getServerContext(), c, table);
+      assertFalse(files.isEmpty());
+      c.tableOperations().compact(table, new CompactionConfig());
```
[the remark this answers is not in the document]
Comment (dlmarion) [reply test/src/main/java/org/apache/accumulo/test/functional/GarbageCollectorTrashEnabledIT.java:77]: it defaults to true

File: test/src/main/java/org/apache/accumulo/test/functional/GarbageCollectorTrashEnabledIT.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+      tid = TableId.of(c.tableOperations().tableIdMap().get(table));
+    }
+    Thread.sleep(10000);
+    assertEquals(files.size(), super.countFilesInTrash(fs, tid));
```
[the remark this answers is not in the document]
Comment (dlmarion) [reply test/src/main/java/org/apache/accumulo/test/functional/GarbageCollectorTrashEnabledIT.java]: Addressed in 5ec3ca8.

File: test/src/main/java/org/apache/accumulo/test/functional/GarbageCollectorTrashDisabledIT.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    super.makeTrashDir(fs);
+    ArrayList<StoredTabletFile> files = null;
+    TableId tid = null;
+    try (AccumuloClient c = Accumulo.newClient().from(getClientProperties()).build()) {
+      files = super.loadData(super.getServerContext(), c, table);
+      assertFalse(files.isEmpty());
+      c.tableOperations().compact(table, new CompactionConfig());
+      tid = TableId.of(c.tableOperations().tableIdMap().get(table));
+    }
+    Thread.sleep(10000);
```
[the remark this answers is not in the document]
Comment (dlmarion) [reply test/src/main/java/org/apache/accumulo/test/functional/GarbageCollectorTrashDisabledIT.java]: Addressed in 5ec3ca8.

File: test/src/main/java/org/apache/accumulo/test/functional/GarbageCollectorTrashDefaultIT.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+      ArrayList<StoredTabletFile> files = super.loadData(super.getServerContext(), c, table);
+      assertFalse(files.isEmpty());
+      c.tableOperations().compact(table, new CompactionConfig());
+      tid = TableId.of(c.tableOperations().tableIdMap().get(table));
+    }
+
+    // The default value for fs.trash.interval is 0, which means that
+    // trash is disabled in the Hadoop configuration. Enabling trash in
+    // Accumulo (GC_TRASH_IGNORE = false) still requires enabling trash in Hadoop
+    Thread.sleep(10000);
```
[the remark this answers is not in the document]
Comment (dlmarion) [reply test/src/main/java/org/apache/accumulo/test/functional/GarbageCollectorTrashDefaultIT.java]: Addressed in 5ec3ca8.

File: test/src/main/java/org/apache/accumulo/test/functional/GarbageCollectorTrashDefaultIT.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+public class GarbageCollectorTrashDefaultIT extends GarbageCollectorTrashBase {
+
+  @Override
+  protected Duration defaultTimeout() {
+    return Duration.ofMinutes(5);
+  }
+
+  @Override
+  public void configure(MiniAccumuloConfigImpl cfg, Configuration hadoopCoreSite) {
+    // By default Hadoop trash is disabled - fs.trash.interval defaults to 0
```
[the remark this answers is not in the document]
Comment (dlmarion) [reply test/src/main/java/org/apache/accumulo/test/functional/GarbageCollectorTrashDefaultIT.java:50]: Addressed in 5ec3ca8.

File: test/src/main/java/org/apache/accumulo/test/functional/GarbageCollectorTrashEnabledCustomPolicyIT.java (hunk 1 above)
Comment (dlmarion) [reply test/src/main/java/org/apache/accumulo/test/functional/GarbageCollectorTrashEnabledCustomPolicyIT.java:52]: Addressed in 5ec3ca8.

File: test/src/main/java/org/apache/accumulo/test/functional/GarbageCollectorTrashEnabledCustomPolicyIT.java [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    try (AccumuloClient c = Accumulo.newClient().from(getClientProperties()).build()) {
+      ReadWriteIT.ingest(c, 10, 10, 10, 0, table);
+      c.tableOperations().flush(table);
+      ArrayList<StoredTabletFile> files1 = getFilesForTable(super.getServerContext(), c, table);
+      assertFalse(files1.isEmpty());
+      assertTrue(files1.stream().allMatch(stf -> stf.getPath().getName().startsWith("F")));
+      c.tableOperations().compact(table, new CompactionConfig());
+      super.waitForFilesToBeGCd(files1);
+      ArrayList<StoredTabletFile> files2 = getFilesForTable(super.getServerContext(), c, table);
+      assertTrue(files2.stream().noneMatch(stf -> stf.getPath().getName().startsWith("F")));
```
Comment (keith-turner) [inline test/src/main/java/org/apache/accumulo/test/functional/GarbageCollectorTrashEnabledCustomPolicyIT.java:99]: A check like the one above would be good here.  Could the not empty check be more specific, do you know how many files to expect?

```suggestion
      assertFalse(files2.isEmpty());
      assertTrue(files2.stream().noneMatch(stf -> stf.getPath().getName().startsWith("F")));
```

**Review by keith-turner — APPROVED** [review]

File: test/src/main/java/org/apache/accumulo/test/functional/GarbageCollectorTrashBase.java [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+import org.apache.accumulo.test.util.Wait;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.LocatedFileStatus;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.RemoteIterator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class GarbageCollectorTrashBase extends ConfigurableMacBase {
```
Comment (ctubbsii) [inline test/src/main/java/org/apache/accumulo/test/functional/GarbageCollectorTrashBase.java:46]:
```suggestion
// base class for ITs that test our legacy trash flag and adoop's trash policy with accumulo-gc
public class GarbageCollectorTrashBase extends ConfigurableMacBase {
```

**Review by ctubbsii — CHANGES_REQUESTED** [review]
Minor changes suggested to add comments and expand the log messages with some helpful details. I was in a rush and didn't really proofread them, so they can probably be worded better than my attempt. Mainly, I just think it's helpful to have a brief description, rather than rely on the name alone to try to figure out the differences between the test cases in the separate classes.

It does strike me that most of the ITs are basically just testing Hadoop's own behavior, rather than our own. Once our prop is removed, these tests can be simplified. I'm not sure we need to test Hadoop behavior so comprehensively in Accumulo, but for now it makes sense, because of the testing of the interaction with our config prop.

File: server/gc/src/test/java/org/apache/accumulo/gc/SimpleGarbageCollectorTest.java [hunk 9]
```diff
@@ -65,6 +65,8 @@ public class SimpleGarbageCollectorTest {
   private SimpleGarbageCollector gc;
   private ConfigurationCopy systemConfig;
   private static SiteConfiguration siteConfig = SiteConfiguration.empty().build();
+  @SuppressWarnings("removal")
+  private final Property p = Property.GC_TRASH_IGNORE;
```
[the remark this answers is not in the document]
Comment (dlmarion) [reply server/gc/src/test/java/org/apache/accumulo/gc/SimpleGarbageCollectorTest.java]: Implemented in 2aa689f

File: server/gc/src/test/java/org/apache/accumulo/gc/SimpleGarbageCollectorTest.java [hunk 10]
```diff
@@ -97,7 +99,7 @@ private ConfigurationCopy createSystemConfig() {
     conf.put(Property.GC_CYCLE_START.getKey(), "1");
     conf.put(Property.GC_CYCLE_DELAY.getKey(), "20");
     conf.put(Property.GC_DELETE_THREADS.getKey(), "2");
-    conf.put(Property.GC_TRASH_IGNORE.getKey(), "false");
+    conf.put(p.getKey(), "false");
```
[the remark this answers is not in the document]
Comment (dlmarion) [reply server/gc/src/test/java/org/apache/accumulo/gc/SimpleGarbageCollectorTest.java]: Implemented in 2aa689f

File: server/gc/src/main/java/org/apache/accumulo/gc/GCRun.java [hunk 11]
```diff
@@ -463,11 +463,15 @@ boolean inSafeMode() {
...
+      log.trace("Accumulo Trash is disabled.");
       return false;
     }
     try {
-      return fs.moveToTrash(path);
+      boolean success = fs.moveToTrash(path);
+      log.trace("Accumulo Trash enabled, moving to trash succeeded?: {}", success);
+      return success;
     } catch (FileNotFoundException ex) {
+      log.error("Error moving to trash", ex);
```
[the remark this answers is not in the document]
Comment (dlmarion) [reply server/gc/src/main/java/org/apache/accumulo/gc/GCRun.java]: Implemented in 2aa689f

File: test/src/main/java/org/apache/accumulo/test/functional/GarbageCollectorTrashBase.java
[the remark this answers is not in the document]
Comment (dlmarion) [reply test/src/main/java/org/apache/accumulo/test/functional/GarbageCollectorTrashBase.java:1]: Implemented in 2aa689f

File: test/src/main/java/org/apache/accumulo/test/functional/GarbageCollectorTrashDisabledIT.java [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+import org.apache.accumulo.core.conf.Property;
+import org.apache.accumulo.core.data.TableId;
+import org.apache.accumulo.core.metadata.StoredTabletFile;
+import org.apache.accumulo.miniclusterImpl.MiniAccumuloConfigImpl;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.CommonConfigurationKeysPublic;
+import org.apache.hadoop.fs.FileSystem;
+import org.junit.jupiter.api.Test;
+
+public class GarbageCollectorTrashDisabledIT extends GarbageCollectorTrashBase {
```
[the remark this answers is not in the document]
Comment (dlmarion) [reply test/src/main/java/org/apache/accumulo/test/functional/GarbageCollectorTrashDisabledIT.java:43]: Implemented in 2aa689f

File: test/src/main/java/org/apache/accumulo/test/functional/GarbageCollectorTrashEnabledCustomPolicyIT.java [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+import org.apache.accumulo.core.metadata.StoredTabletFile;
+import org.apache.accumulo.miniclusterImpl.MiniAccumuloConfigImpl;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.CommonConfigurationKeysPublic;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.TrashPolicyDefault;
+import org.junit.jupiter.api.Test;
+
+public class GarbageCollectorTrashEnabledCustomPolicyIT extends GarbageCollectorTrashBase {
```
[the remark this answers is not in the document]
Comment (dlmarion) [reply test/src/main/java/org/apache/accumulo/test/functional/GarbageCollectorTrashEnabledCustomPolicyIT.java:47]: Implemented in 2aa689f

File: test/src/main/java/org/apache/accumulo/test/functional/GarbageCollectorTrashEnabledIT.java [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+import org.apache.accumulo.core.conf.Property;
+import org.apache.accumulo.core.data.TableId;
+import org.apache.accumulo.core.metadata.StoredTabletFile;
+import org.apache.accumulo.miniclusterImpl.MiniAccumuloConfigImpl;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.CommonConfigurationKeysPublic;
+import org.apache.hadoop.fs.FileSystem;
+import org.junit.jupiter.api.Test;
+
+public class GarbageCollectorTrashEnabledIT extends GarbageCollectorTrashBase {
```
[the remark this answers is not in the document]
Comment (dlmarion) [reply test/src/main/java/org/apache/accumulo/test/functional/GarbageCollectorTrashEnabledIT.java:43]: Implemented in 2aa689f

**Comment by EdColeman** [comment]
An argument to test the hadoop functionality - at least to some extent is to validate the the function continues to work as expected in future releases - if the "hadoop" functionality changed (or broke) we may catch that in out automated test runs and then we would be in a position to take some action - maybe we need to do something different? Submit a patch to hadoop?  Or at a minimum, warn users that the function may not work as expected after x.y.z.

I'm not sure how extensively we are checking the behavior here, but some validation might be valuable.

**Comment by ctubbsii** [comment]
> An argument to test the hadoop functionality - at least to some extent is to validate the the function continues to work as expected in future releases - if the "hadoop" functionality changed (or broke) we may catch that in out automated test runs and then we would be in a position to take some action - maybe we need to do something different? Submit a patch to hadoop? Or at a minimum, warn users that the function may not work as expected after x.y.z.
> 
> I'm not sure how extensively we are checking the behavior here, but some validation might be valuable.

My argument against, as a general rule, is that regression testing for Hadoop features should be contributed to Hadoop. We cannot be the landing zone for testing all the Hadoop features we think Hadoop is not testing adequately. I suggest contributing those tests to Hadoop to make Hadoop better, and watching its build statuses/change history for the features we care about, and try to focus our tests on testing our code.

I'm definitely not arguing against testing... just advocating for contributing them to the proper venue. We shouldn't treat Accumulo as a personal playground for testing other libraries we care about. We should play more nicely with those other open source projects, and contribute tests to them, if we think they need that testing.

That said, everything is on a case-by-case basis, and I'm not sure exactly what, if anything, should change here once the property is dropped, with respect to this testing, until I look at it, which I will do shortly. But I think it's important to stake out a general principle, so we don't suffer from scope creep.
